### PR TITLE
fix cancel render loop

### PR DESCRIPTION
**Pull request by sebavan**
https://forum.babylonjs.com/t/stopping-render-loop-and-then-immediately-running-it-again-causes-double-frames/48599

### CODE REVIEW

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14868/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14868/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14868/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14868/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14868/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14868/merge#BCU1XR#0

File: packages/dev/core/src/Engines/engine.ts [hunk 1]
```diff
@@ -1275,7 +1279,7 @@ export class Engine extends ThinEngine {
             }
         }
 
-        if (this._activeRenderLoops.length > 0) {
+        if (this._activeRenderLoops.length > 0 && this._frameHandler === -1) {
```
Comment (RaananW) [inline packages/dev/core/src/Engines/engine.ts:1282]: In what case is not -1?

**Review by RaananW — APPROVED** [review]
LGTM. 

I thought it might not be needed to check for frameHandler > -1 when cancelling, but better stay safe there.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgpuplaywright/index.html

File: packages/dev/core/src/Engines/engine.ts (hunk 1 above)
Comment (sebavan) [reply packages/dev/core/src/Engines/engine.ts:1282]: if during the frame you call cancel then runRenderLoop again with a different function for instance. This is indeed the main issue discovered on the associated forum thread.

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgl2playwright/index.html

**Review by deltakosh — APPROVED** [review]

File: packages/dev/core/src/Engines/engine.ts [hunk 2]
```diff
@@ -1242,18 +1242,22 @@ export class Engine extends ThinEngine {
     }
 
     protected _cancelFrame() {
-        if (this._renderingQueueLaunched && this.customAnimationFrameRequester) {
-            this._renderingQueueLaunched = false;
-            const { cancelAnimationFrame } = this.customAnimationFrameRequester;
-            if (cancelAnimationFrame) {
-                cancelAnimationFrame(this.customAnimationFrameRequester.requestID);
+        if (this.customAnimationFrameRequester) {
+            if (this._frameHandler > -1) {
```
Comment (Popov72) [inline packages/dev/core/src/Engines/engine.ts:1246]:
```suggestion
            if (this._frameHandler !== 0) {
```
The [requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) doc says that the returned value is a non-zero long value and that no other assumption should be made.

File: packages/dev/core/src/Engines/engine.ts [hunk 3]
```diff
@@ -1242,18 +1242,22 @@ export class Engine extends ThinEngine {
...
 
     protected _cancelFrame() {
-        if (this._renderingQueueLaunched && this.customAnimationFrameRequester) {
-            this._renderingQueueLaunched = false;
-            const { cancelAnimationFrame } = this.customAnimationFrameRequester;
-            if (cancelAnimationFrame) {
-                cancelAnimationFrame(this.customAnimationFrameRequester.requestID);
+        if (this.customAnimationFrameRequester) {
+            if (this._frameHandler > -1) {
+                this._frameHandler = -1;
```
Comment (Popov72) [inline packages/dev/core/src/Engines/engine.ts:1247]:
```suggestion
                this._frameHandler = 0;
```

File: packages/dev/core/src/Engines/engine.ts [hunk 4]
```diff
@@ -1242,18 +1242,22 @@ export class Engine extends ThinEngine {
...
+                    cancelAnimationFrame(this.customAnimationFrameRequester.requestID);
+                }
             }
         } else {
             super._cancelFrame();
         }
     }
 
     public _renderLoop(): void {
+        this._frameHandler = -1;
```
Comment (Popov72) [inline packages/dev/core/src/Engines/engine.ts:1259]:
```suggestion
        this._frameHandler = 0;
```

File: packages/dev/core/src/Engines/engine.ts (hunk 1 above)
Comment (Popov72) [inline packages/dev/core/src/Engines/engine.ts:1282]:
```suggestion
        if (this._activeRenderLoops.length > 0 && this._frameHandler === 0) {
```

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 5]
```diff
@@ -582,7 +581,7 @@ export class ThinEngine {
     private _emptyTexture2DArray: Nullable<InternalTexture>;
 
     /** @internal */
-    public _frameHandler: number;
+    public _frameHandler: number = -1;
```
Comment (Popov72) [inline packages/dev/core/src/Engines/thinEngine.ts:584]:
```suggestion
    public _frameHandler: number = 0;
```

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 6]
```diff
@@ -1586,24 +1585,28 @@ export class ThinEngine {
     }
 
     protected _cancelFrame() {
-        if (this._renderingQueueLaunched && this._frameHandler) {
-            this._renderingQueueLaunched = false;
+        if (this._frameHandler > -1) {
```
Comment (Popov72) [inline packages/dev/core/src/Engines/thinEngine.ts:1588]:
```suggestion
        if (this._frameHandler !== 0) {
```

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 7]
```diff
@@ -1586,24 +1585,28 @@ export class ThinEngine {
     }
 
     protected _cancelFrame() {
-        if (this._renderingQueueLaunched && this._frameHandler) {
-            this._renderingQueueLaunched = false;
+        if (this._frameHandler > -1) {
+            const handlerToCancel = this._frameHandler;
+            this._frameHandler = -1;
```
Comment (Popov72) [inline packages/dev/core/src/Engines/thinEngine.ts:1590]:
```suggestion
            this._frameHandler = 0;
```

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 8]
```diff
@@ -1586,24 +1585,28 @@ export class ThinEngine {
...
                 }
             }
-            return clearTimeout(this._frameHandler);
+            return clearTimeout(handlerToCancel);
         }
     }
 
     /** @internal */
     public _renderLoop(): void {
+        this._frameHandler = -1;
```
Comment (Popov72) [inline packages/dev/core/src/Engines/thinEngine.ts:1608]:
```suggestion
        this._frameHandler = 0;
```

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 9]
```diff
@@ -1625,10 +1628,8 @@ export class ThinEngine {
             }
         }
 
-        if (this._activeRenderLoops.length > 0) {
+        if (this._activeRenderLoops.length > 0 && this._frameHandler === -1) {
```
Comment (Popov72) [inline packages/dev/core/src/Engines/thinEngine.ts:1631]:
```suggestion
        if (this._activeRenderLoops.length > 0 && this._frameHandler === 0) {
```

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 10]
```diff
@@ -1717,9 +1718,8 @@ export class ThinEngine {
 
         this._activeRenderLoops.push(renderFunction);
 
-        if (!this._renderingQueueLaunched) {
-            this._renderingQueueLaunched = true;
-            this._boundRenderFunction = () => this._renderLoop();
+        // On the first added function, start the render loop.
+        if (this._activeRenderLoops.length === 1 && this._frameHandler === -1) {
```
Comment (Popov72) [inline packages/dev/core/src/Engines/thinEngine.ts:1722]: I think you don't need the `this._activeRenderLoops.length === 1` part? If `this._frameHandler == -1`, it means no RAF is in flight.

**Review by Popov72 — CHANGES_REQUESTED** [review]

File: packages/dev/core/src/Engines/thinEngine.ts (hunk 10 above)
Comment (sebavan) [reply packages/dev/core/src/Engines/thinEngine.ts:1722]: I was concerned by 
image

For back compat but I suppose I should make the list readonly so we can get read of the check ?

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14868/merge/testResults/webgpuplaywright/index.html

**Review by Popov72 — APPROVED** [review]

**Pull request merged**